File: C++/563.cpp
__________________________________________________________________________________________________
sample 20 ms submission
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
static int speed = [](){ios::sync_with_stdio(0);cin.tie(0);cout.tie(0);return 0;}();
class Solution {
public:
    int tilt=0;
    int tiltTree(TreeNode *root) {
        if (!root) return 0;
        int left = tiltTree(root->left);
        int right = tiltTree(root->right);
        tilt+=abs(left-right);
        return left+right+root->val;
    }
    int findTilt(TreeNode* root) {
        tiltTree(root);
        return tilt;
    }
};
__________________________________________________________________________________________________
sample 21904 kb submission
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
void solve(TreeNode* root,int &sum)
{
    if(root == NULL)
        return;
    if(root->left == NULL && root->right == NULL)
    {
        return;
    }
    
    solve(root->left,sum);
    solve(root->right,sum);
    
    int s1 = 0;
    int s2 = 0;
    if(root->left != NULL)
        s1 = root->left->val;
    if(root->right != NULL)
        s2 = root->right->val;
    
    sum += abs(s1-s2);
    root->val += s1 + s2;
    
    return;
}

class Solution {
public:
    int findTilt(TreeNode* root) {
        
        if(root == NULL)
            return 0;
    
        int sum = 0;
        solve(root,sum);        
        
        return sum;
    }
};
__________________________________________________________________________________________________

File: Java/563.java
__________________________________________________________________________________________________
sample 0 ms submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int tilt = 0;
    public int findTilt(TreeNode root) {
        dfs(root);
        return tilt;
    }
    
    private int dfs(TreeNode node) {
        if(node == null) {
            return 0;
        }
        if(node.left == null && node.right == null) {
            return node.val;
        }
        int sum = 0, valueLeft = 0, valueRight = 0;
        valueLeft += dfs(node.left);
        valueRight += dfs(node.right);
        tilt += valueRight > valueLeft ? valueRight - valueLeft : valueLeft - valueRight;
        sum = valueLeft + valueRight + node.val;
        return sum;
        
    }
}
__________________________________________________________________________________________________
sample 37520 kb submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int tilt = 0;
    
    public int findTilt(TreeNode root) {
        helper(root);
        return tilt;
    }
    
    private int helper(TreeNode root) {
        if (root == null)
            return 0;
        int left = helper(root.left);
        int right = helper(root.right);
        tilt += Math.abs(left - right);
        return left + right + root.val;
    }
}
__________________________________________________________________________________________________
